[1.18.x] Allow using DeferredRegisters for vanilla registries

SizableShrimp opened the pull request:
This PR expands the functionality of `DeferredRegister`s and `RegistryObject`s to be able to use them for vanilla registries which have no associated forge registry.
Unlike #8495, this PR contains no binary or source compatibility breaks and should therefore be only a benefit to modders.

To use a `DeferredRegister` for a vanilla registry, use the `create` factory that takes a `ResourceKey`. Example:
```java
public static final DeferredRegister<RecipeType<?>> RECIPE_TYPES = DeferredRegister.create(Registry.RECIPE_TYPE_REGISTRY, MODID);
```
Then, register objects like you would with any normal `DeferredRegister`. Example:
```java
    public static final RegistryObject<RecipeType<?>> TEST_RECIPE_TYPE = RECIPE_TYPES.register("test", () -> new RecipeType<>() {});
```
This example has also been added to `DeferredRegistryTest`.

This should cleanup the admittedly ugly previous solution of "just sticking it wherever" and using `Lazy.of()` or static non-final fields that start out as null, then registering from a random `RegistryEvent.Register<?>`.
More testing would be appreciated, e.g. worldgen people.

SizableShrimp committed:
Allow using DeferredRegisters for vanilla regs

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Add JS coremod for `RegistryObject` binary compat

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

KnightMiner commented on src/main/java/net/minecraftforge/registries/VanillaRegisterEvent.java:
```diff
@@ -0,0 +1,22 @@
+import net.minecraftforge.fml.event.IModBusEvent;
+import org.jetbrains.annotations.NotNull;
+
+class VanillaRegisterEvent extends Event implements IModBusEvent
```
Any reason this is package protected instead of public? An event like this would be useful outside deferred registers for registering things into vanilla registries using the classic event style.

Also, could this be made a generic event? That way you could subscribe to `VanillaRegisterEvent<RecipeType<?>` for instance.

LexManos replied:
The entire concept is that we can no longer garentee the type. The plan is to remove types all together in 1.19.
And the reason it is private is because We do not wish modders to use it. Instead, using DR is the desired API. I even told Shrimp that this shouldn't be a event at all, instead it should be a manual callback on our end. 

In 1.19 the Register event will become non-generic, and instead be keyed on Registry Name.

This event is purely a stop gap until we can make that breaking change.

SizableShrimp committed:
Fire VanillaRegisterEvent for builtin registries

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented:
This PR is currently blocked as both myself and the PR author are experiencing a very strange issue where the `@ObjectHolder`-annotated field in [`MultiLayerModelTest`](https://github.com/SizableShrimp/MinecraftForge/blob/e8ec6b892760f28467b891058bf99b70cbae2fef/src/test/java/net/minecraftforge/debug/client/model/MultiLayerModelTest.java#L30-L31) seems to fail almost at random (with a bias for not working). Somehow, the injection for that field fails seemingly at random, causing a `NullPointerException` when its corresponding `BlockItem` is created and registered afterwards.

Until this issue is debugged and fixed, this PR cannot move forward to assignment. We appreciate it if others try this PR locally to see if they experience this issue as well.

SizableShrimp committed:
Fix missed RegistryAccess patch

VanillaRegisterEvent fires for dynamic registries
Deleted hunk that is no longer needed - BuiltinRegistries.ACCESS already wraps around BuiltinRegistries.REGISTRY

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix registering to builtin registries

Clarified comments in DeferredRegistryTest and fix test
Synchronize additions and removals of object holders

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -111,15 +115,25 @@ private RegistryObject() {
+                        LOGGER.error(
+                                "Unable to find registry for type " + baseType.getName() + " for mod \"" + modid + "\". Check the 'caused by' to see further stack.",
+                                callerStack
+                        );
```
Use logger parameter substitution, and move this all to one line:
```suggestion
                        LOGGER.error("Unable to find registry for type {} for mod \"{}\". Check the 'caused by' to see further stack.", baseType.getName(), modid, callerStack);
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -140,16 +154,24 @@ private RegistryObject(final ResourceLocation name, final ResourceLocation regis
+                    {
+                        invalidRegistry = true;
+                        LOGGER.error(
                                 "Unable to find registry with key " + registryName + " for mod \"" + modid + "\". Check the 'caused by' to see further stack.",
```
Move to logger parameter substitution, and move to one line (see previous comment).

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/VanillaRegisterEvent.java:
```diff
@@ -0,0 +1,22 @@
+import net.minecraftforge.fml.event.IModBusEvent;
+import org.jetbrains.annotations.NotNull;
+
+class VanillaRegisterEvent extends Event implements IModBusEvent
```
While this event is a stopgap measure as said by @LexManos, I feel that it should still be documented stating the same, along with referencing and linking to `DeferredRegister`. It may also be worth considering (terminally) deprecating this, to definitely mark this for deletion come 1.19.

sciwhiz12 commented on src/main/resources/coremods/registry_object_binary_compat.js:
```diff
@@ -0,0 +1,54 @@
+var InsnNode = Java.type('org.objectweb.asm.tree.InsnNode')
+
+// TODO 1.19: Delete this whole coremod. We won't need to maintain binary compatibility then.
+function initializeCoreMod() {
```
Even though this will be gone in 1.19, please still add comments explaining what this coremod does, to avoid having to read through the coremod's logic. (Some may say that is unnecessary, but I say that if we're adding something here that not many understand, then it's a good place to add comments.)

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -90,34 +90,32 @@
 
-    private final ResourceLocation registryName;
-    private final Class<T> superType;
+    private final ResourceKey<? extends Registry<T>> registryKey;
+    private final Class<? extends IForgeRegistryEntry<?>> superType;
```
These should be marked as `@Nullable`. Additionally, add comments which denote which of the two are nullable at which time.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -90,34 +90,32 @@
-    private IForgeRegistry<T> type;
-    private Supplier<IForgeRegistry<T>> typeSupplier;
-    private Supplier<RegistryBuilder<T>> registryFactory;
+    private Supplier<RegistryBuilder<?>> registryFactory;
     private SetMultimap<TagKey<T>, Supplier<T>> optionalTags;
```
These should be marked as `@Nullable`.

sciwhiz12 commented on patches/minecraft/net/minecraft/core/RegistryAccess.java.patch:
```diff
@@ -1,11 +1,10 @@
+@@ -186,6 +_,7 @@
        ResourceKey<? extends Registry<E>> resourcekey = p_211083_.f_123101_();
        Registry<E> registry = BuiltinRegistries.f_206379_.m_175515_(resourcekey);
-+      if (!resourcekey.equals(Registry.f_122818_))
-+         registry = ((Registry<Registry<E>>)BuiltinRegistries.f_123858_).m_6246_((ResourceKey<Registry<E>>)resourcekey);
```
Why was this patch deleted? (Asking so it's documented here for future reference.)

SizableShrimp replied:
Per my commit description:
> Deleted hunk that is no longer needed - BuiltinRegistries.ACCESS already wraps around BuiltinRegistries.REGISTRY

sciwhiz12 commented:
The above noted issue was resolved. To my understanding, the issue was caused by three reasons:

  1. Because the `PlacedFeature` registry did not exist in the root (regular) registry at `Registry.REGISTRY` (and neither exists as a Forge-added registry), the handler registered by `RegistryObject` to `ObjectHolderRegistry` threw an `IllegalStateException`.

  2. The above noted exception was swallowed because of the internals of FML.[^1] However, the exception still would interrupt `ObjectHolderRegistry`'s calling of its internal collection. If the `RegistryObject` was called (and threw) before the `ObjectHolder`s, then this issue happens.

  3. `ObjectHolderRegistry` uses a `HashSet` for its internal collection of callbacks. The `HashSet`'s ordering is unspecified and may change between runs of the JVM, causing the random behavior observed earlier.

The issue was mitigated by changing the exception into a simple `Logger` statement. While this increases the chance of `RegistryObject`s being created for invalid registries, it is an acceptable compromise given that mod developer should be testing their mods and catching these errors themselves, as well as the daunting task of cleaning up and refactoring FML (which is out-of-scope for this PR).

[^1]: FML uses `CompletableFuture<List<Throwable>>`s for the various tasks and hooks during the state transition process, where the contained list value seemed to be originally meant as the list of exceptions thrown during the task or hook. 
However, investigations conclude that later refactors of the state transition code eliminated that use of the list but, for unknown reasons, was kept in as the type parameter to `CompletableFuture`. 
Furthermore, the registry event dispatch hooks kept using the list as it was originally intended, handling any exceptions thrown during the hook into the list. Because the list is never queried in the code, this had the effect of discarding the exception.

SizableShrimp committed:
Update based on comments by sciwhiz12

Added some nullability annotations
Added some documentation
Fixed 2 logger calls

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Reinstate enchanting_fuels as an optional tag

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Fix VanillaRegisterEvent for builtin registries

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/test/java/net/minecraftforge/debug/DeferredRegistryTest.java:
```diff
@@ -57,7 +71,19 @@ public DeferredRegistryTest() {
+        MinecraftForge.EVENT_BUS.addListener(this::serverStarted);
+    }
+
+    public void serverStarted(ServerStartedEvent event) {
```
Formatting: braces on newlines.
```suggestion
    public void serverStarted(ServerStartedEvent event)
    {
```

SizableShrimp committed:
Fix formatting in DeferredRegistryTest

Co-authored-by: sciwhiz12 <sciwhiz12@sciwhiz12.tk>

SizableShrimp committed:
Revert "Reinstate enchanting_fuels as an optional tag"

This reverts commit ddd6fc388cdb192b3467fd7f7d716cbdb44e2fe8.

SizableShrimp committed:
Add concept of optional-registry RegistryObjects

Expand documentation
Fix filling object holders for vanilla and builtin registries
Catch and aggregate exceptions when applying object holders

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
Change optionalRegistry to #optional method

Add dummy deferred register to DeferredRegistryTest testing createOptional

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/DeferredRegister.java:
```diff
@@ -68,38 +77,89 @@
-     * DeferredRegister factory for custom registries that are made during the {@link NewRegistryEvent} event
-     * or with {@link #makeRegistry(Class, Supplier)}.
+     * DeferredRegister factory for custom forge registries, {@link Registry vanilla registries},
+     * or {@link BuiltinRegistries built-in registries}.
```
For clarity reasons, the first sentence of the javadocs for the following four methods should distinguish each method from each other (as this is used as the "summary" of the method in javadoc).

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ForgeRegistries.java:
```diff
@@ -147,12 +145,13 @@ private static void init() {}
 
     /**
-     * This function is just to make sure static initializers in other classes have run and setup their registries before we query them.
+     * This function is just to make sure static inializers in other classes have run and setup their registries before we query them.
```
Could you keep the typo correction?

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/ObjectHolderRegistry.java:
```diff
@@ -171,7 +171,23 @@ public static void applyObjectHolders()
+            }
+        });
+
+        if (aggregate.getSuppressed().length > 0) {
```
Formatting: braces on newlines.
```suggestion
        if (aggregate.getSuppressed().length > 0)
        {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -51,27 +56,75 @@
 
     /**
-     * Factory for a {@link RegistryObject} that stores the value of an object in a registry once it is ready.
+     * Factory for a {@link RegistryObject} that stores the value of an object from a registry once it is ready.
```
The first sentence should differentiate this factory from the other factories, rather than being uniform.

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -51,27 +56,75 @@
+     * @see #of(ResourceLocation, IForgeRegistry)
+     * @see #of(ResourceLocation, ResourceLocation, String)
+     */
+    public static <T, U extends T> RegistryObject<U> optional(final ResourceLocation name, final ResourceKey<? extends Registry<T>> registryKey, String modid) {
```
First, `ofOptional` would probably be more conforming to the current conventional names of the factories (akin to `Optional.of` and `Optional.ofNullable`).
Formatting: braces on newlines.
```suggestion
    public static <T, U extends T> RegistryObject<U> ofOptional(final ResourceLocation name, final ResourceKey<? extends Registry<T>> registryKey, String modid)
    {
```

sciwhiz12 commented on src/main/java/net/minecraftforge/registries/RegistryObject.java:
```diff
@@ -51,27 +56,75 @@
+     * @see #of(ResourceLocation, IForgeRegistry)
+     * @see #of(ResourceLocation, ResourceKey, String)
+     */
+    public static <T, U extends T> RegistryObject<U> optional(final ResourceLocation name, final ResourceLocation registryName, String modid) {
```
Same comment for `ofOptional`. Formatting: braces on newlines.
```suggestion
    public static <T, U extends T> RegistryObject<U> optional(final ResourceLocation name, final ResourceLocation registryName, String modid)
    {
```

SizableShrimp committed:
Cleanup docs in DeferredRegister and RegistryObject

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>

SizableShrimp committed:
of -> create, optional -> createOptional

Signed-off-by: SizableShrimp <sizableshrimp@sizableshrimp.me>
